rosbag2/src/rosbag2/sequential_reader.cpp

```diff
@@ -12,88 +13,50 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-#include "rosbag2/sequential_reader.hpp"
-
 #include <memory>
-#include <stdexcept>
-#include <string>
 #include <utility>
 #include <vector>
 
 #include "rosbag2/info.hpp"
+#include "rosbag2/sequential_reader.hpp"
+
+#include "rosbag2/impl/sequential_reader_impl.hpp"
 
 namespace rosbag2
 {
 
 SequentialReader::SequentialReader(
   std::unique_ptr<rosbag2_storage::StorageFactoryInterface> storage_factory,
   std::shared_ptr<SerializationFormatConverterFactoryInterface> converter_factory)
-: storage_factory_(std::move(storage_factory)), converter_factory_(std::move(converter_factory)),
-  converter_(nullptr)
+: reader_impl_(std::make_unique<SequentialReaderImpl>(
```

Comment:
I feel like this should actually be an input parameter. The way I feel composition should work is that the constructor takes its specialization as parameters.
And with the `reader_impl_` being a pointer to an interface class, this can be changed dynamically. 

The way I see this:

```
SequentialReader::SequentialReader(
  ... // storage,
  ... // converter,
  std::unique_ptr<BaseReaderInterface> reader_impl)
: reader_impl_(std::move(reader_impl))
{}
```

This way a SequentialReader can easily be extended to take in an implementation of the `BaseReaderInterface` which allows e.g. compression or future enhancements. 
On top of that the SequentialReader class can be final.

In fact, with this pattern the `SequentialReader` could be renamed to solely `Reader` where the input parameter dictates its behavior. So another implementation of the `BaseReaderInterface` could be a non sequential reading.